### ISSUE TITLE
Error: Could not find unallocated temp file

### ISSUE THREAD

**Issue by AndreasPK** [haskell/hsc2hs#33]
I'm hitting this error repeatedly.

As I understand it this is a side effect of wonky code to find a temp file in `withTempFile`.  
The code is there since  `openTempFile` is not atomic pre GHC-8.6.

The easy solution would be to just use openTempFile, but this does not work with older GHCs.

So if we are willing to pay a price would calling `GetTempFileName` via FFI and replicating the GHC logic be a reasonable solution?

**Comment by AndreasPK** [comment]
For more context, I hit this error when building GHC HEAD.

```
| Run Hsc2Hs Stage1: libraries/base/System/Environment/ExecutablePath.hsc => _prof/stage1/libraries/base/build/System/Environment/ExecutablePath.hs
Could not find unallocated temp file
Error when running Shake build system:
  at action, called at src\Rules.hs:71:19 in main:Rules
  at need, called at src\Rules.hs:93:5 in main:Rules
* Depends on: _prof/stage1/lib/package.conf.d/base-4.13.0.0.conf
  at need, called at src\Rules\Register.hs:113:5 in main:Rules.Register
* Depends on: _prof/stage1/libraries/base/build/libHSbase-4.13.0.0.a
  at need, called at src\Rules\Library.hs:144:5 in main:Rules.Library
* Depends on: _prof/stage1/libraries/base/build/GHC/IO/Device.o
  at &%>, called at src\Rules\Compile.hs:47:9 in main:Rules.Compile
* Depends on: _prof/stage1/libraries/base/build/GHC/IO/Device.o _prof/stage1/libraries/base/build/GHC/IO/Device.hi
  at apply1, called at src\Development\Shake\Internal\Rules\Oracle.hs:159:32 in shake-0.18.3-86a4ec57ddae5ecfda9c7c523ad27c4c920b10a1:Development.Shake.Internal.Rules.Oracle
* Depends on: OracleQ (KeyValues ("_prof/stage1/libraries/base/.dependencies","_prof/stage1/libraries/base/build/GHC/IO/Device.o"))
  at need, called at src\Hadrian\Oracles\TextFile.hs:96:9 in main:Hadrian.Oracles.TextFile
* Depends on: _prof/stage1/libraries/base/.dependencies
  at readFile', called at src\Rules\Dependencies.hs:34:19 in main:Rules.Dependencies
  at need, called at src\Development\Shake\Internal\Derived.hs:118:15 in shake-0.18.3-86a4ec57ddae5ecfda9c7c523ad27c4c920b10a1:Development.Shake.Internal.Derived
* Depends on: _prof/stage1/libraries/base/.dependencies.mk
  at need, called at src\Rules\Dependencies.hs:26:9 in main:Rules.Dependencies
* Depends on: _prof/stage1/libraries/base/build/System/Environment/ExecutablePath.hs
* Raised the exception:
user error (Development.Shake.cmd, system command failed
Command line: _prof/stage0/bin/hsc2hs.exe --cc=E:/ghc_t17004/inplace/mingw/bin/gcc.exe --ld=E:/ghc_t17004/inplace/mingw/bin/gcc.exe --cflag=-Iincludes --cflag=-I_prof/generated --cflag=-I_prof/stage1/libraries/base/build --cflag=-I_prof/stage1/libraries/base/build/include --cflag=-
Ilibraries/base/include --cflag=-IE:/ghc_t17004/_prof/stage1/lib/x86_64-windows-ghc-8.9.0.20191002/integer-gmp-1.0.2.0/include --cflag=-IE:/ghc_t17004/_prof/stage1/lib/x86_64-windows-ghc-8.9.0.20191002/rts-1.0/include --cflag=-Wall --cflag=-Wno-error=inline --cflag=-include --cflag
=_prof/stage1/libraries/base/build/autogen/cabal_macros.h --cflag=-Dx86_64_HOST_ARCH=1 --cflag=-Dmingw32_HOST_OS=1 --cflag=-D__GLASGOW_HASKELL__=809 --template=E:/ghc_t17004/_prof/stage0/lib/template-hsc.h libraries/base/System/Environment/ExecutablePath.hsc -o _prof/stage1/librari
es/base/build/System/Environment/ExecutablePath.hs
Exit code: 1
Stderr:
Could not find unallocated temp file
)

```

**Comment by RyanGlScott** [comment]
@Mistuke believes that this should be fixed now that #36/#37 have landed. Please re-open if this is not the case.
